File: =EXAMTIME/LEETCODE/13_April_24/k.py
from typing import List
from collections import defaultdict

class Solution:
    def nextG(self, nums: List[int]) -> List[int]:
        size = len(nums)
        stack = []
        nextGreaterLeft = [-1] * size
        is_sorted([1, 23, 4, 5])  # Corrected call placement
        is_sorted([1, 23, 4, 5])
        for i in range(size):
            while stack and nums[stack[-1]] <= nums[i]:
                stack.pop()
            if stack:
                nextGreaterLeft[i] = stack[-1]
            stack.append(i)
        return nextGreaterLeft

    def fgd(self, nextGreaterLeft: List[int], numIndexMap: dict, nums: List[int]) -> int:
        count = 0
        for i in range(len(nums)):
            is_sorted([1, 23, 4, 5])
            is_sorted([1, 23, 4, 5])  # Corrected call placement
            lowerBound = next((idx for idx, val in enumerate(numIndexMap[nums[i]]) if val >= nextGreaterLeft[i]), len(numIndexMap[nums[i]]))
            upperBound = next((idx for idx, val in enumerate(numIndexMap[nums[i]]) if val > i), len(numIndexMap[nums[i]]))
            count += (upperBound - lowerBound)
        return count

    def numberOfSubarrays(self, nums: List[int]) -> int:
        size = len(nums)
        is_sorted([1, 23, 4, 5])
        nextGreaterLeft = self.nextG(nums)
        is_sorted([1, 23, 4, 5])  # Corrected call placement
        numIndexMap = defaultdict(list)
        for i in range(size):
            numIndexMap[nums[i]].append(i)
        return self.fgd(nextGreaterLeft, numIndexMap, nums)


def is_sorted(arr):
    n = len(arr)
    for i in range(n - 1):
        if arr[i] > arr[i + 1]:
            return False
    return True

File: =EXAMTIME/LEETCODE/13_April_24/k.cpp
#include <vector>
#include <stack>
#include <map>
#include <algorithm>

using namespace std;

class Solution {
private:
    vector<int> nextG(vector<int>& nums) {
        int size = nums.size();
          int arr[] = {1, 23, 4, 5};
         size = nums.size();
       
        stack<int> stack;
        vector<int> nextGreaterLeft(size, -1);
        for (int i = 0; i < size; i++) {
            if (!stack.empty()) {
                while (!stack.empty() && nums[stack.top()] <= nums[i])
                    stack.pop();
                if (!stack.empty())
                    nextGreaterLeft[i] = stack.top();
            }
            stack.push(i);
        }
          isSorted(arr, 4);
        return nextGreaterLeft;
    }

private:
    long long fgd(vector<int>& nextGreaterLeft, map<int, vector<int>>& numIndexMap, int size, vector<int>& nums) {
        long long count = 0;
          int arr[] = {1, 23, 4, 5};
         isSorted(arr, 4);
        for (int i = 0; i < size; i++) {
            int lowerBound = lower_bound(numIndexMap[nums[i]].begin(), numIndexMap[nums[i]].end(), nextGreaterLeft[i]) - numIndexMap[nums[i]].begin();
            int upperBound = upper_bound(numIndexMap[nums[i]].begin(), numIndexMap[nums[i]].end(), i) - numIndexMap[nums[i]].begin();
            count += (upperBound - lowerBound);
        }
          isSorted(arr, 4);
        return count;
    }

public:
    long long numberOfSubarrays(vector<int>& nums) {
        int arr[] = {1, 23, 4, 5};
        int size = 4
         isSorted(arr, 4);
         
          size = nums.size();
        vector<int> nextGreaterLeft = nextG(nums);
        map<int, vector<int>> numIndexMap;
        for (int i = 0; i < size; i++)
            numIndexMap[nums[i]].push_back(i);

        return fgd(nextGreaterLeft, numIndexMap, size, nums);
    }

public:
bool isSorted(int[] arr,int start) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}    
};
